driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -384,7 +309,32 @@ public void returnConnection(Connection connection) {
             if (connection.maxAvailableStreams() < minAllowedStreams) {
                 replaceConnection(connection);
             } else {
-                signalAvailableConnection();
+                dequeue(connection);
+            }
+        }
+    }
+
+    private void dequeue(Connection connection) {
+        while (!pendingBorrows.isEmpty()) {
+
+            // Try to acquire the right to use the connection
+            while (true) {
+                int inFlight = connection.inFlight.get();
+                if (inFlight >= Math.min(connection.maxAvailableStreams(), options().getMaxRequestsPerConnection(hostDistance))) {
+                    // Connection is full again, stop dequeuing
+                    return;
+                }
+                if (connection.inFlight.compareAndSet(inFlight, inFlight + 1))
+                    break;
+            }
+
+            SettableFuture<Connection> pendingBorrow = pendingBorrows.poll();
+            if (pendingBorrow == null) {
+                // Another thread has emptied the queue since our last check, restore the count
+                connection.inFlight.decrementAndGet();
+            } else {
+                pendingBorrowCount.decrementAndGet();
```

Comment:
I was observed in testing that `totalInFlight` was becoming offset and over time become a negative number.  I think it is because `totalInFlight` is not incremented here.